verilog/analysis/symbol_table_test.cc

```diff
@@ -2183,6 +2183,395 @@ TEST(BuildSymbolTableTest, ModuleDeclarationLocalsDependOnParameter) {
+  EXPECT_EMPTY_STATUSES(build_diagnostics);
+
+  MUST_ASSIGN_LOOKUP_SYMBOL(module_m, root_symbol, "m");
+  EXPECT_EQ(module_m_info.metatype, SymbolMetaType::kModule);
+  EXPECT_EQ(module_m_info.file_origin, &src);
+  EXPECT_EQ(module_m_info.declared_type.syntax_origin,
+            nullptr);  // there is no module meta-type
+
+  MUST_ASSIGN_LOOKUP_SYMBOL(a_variable, module_m, "a");
+  EXPECT_EQ(a_variable_info.metatype, SymbolMetaType::kDataNetVariableInstance);
+  const ReferenceComponentNode* a_type_ref =
+      a_variable_info.declared_type.user_defined_type;
+  EXPECT_EQ(a_type_ref, nullptr);  // implicit type is primitive type
+  EXPECT_TRUE(a_variable_info.declared_type.implicit);
+
+  EXPECT_EQ(module_m_info.local_references_to_bind.size(), 1);
+  const auto ref_map(module_m_info.LocalReferencesMapViewForTesting());
+
+  ASSIGN_MUST_FIND(a_refs, ref_map, "a");
+  ASSERT_EQ(a_refs.size(), 1);  // all references to "a" parameter
+  for (const auto& a_ref : a_refs) {
+    const ReferenceComponent& a_ref_comp(a_ref->components->Value());
+    EXPECT_EQ(a_ref_comp.ref_type, ReferenceType::kUnqualified);
+    EXPECT_EQ(a_ref_comp.required_metatype, SymbolMetaType::kUnspecified);
+    EXPECT_EQ(a_ref_comp.identifier, "a");
+    EXPECT_EQ(a_ref_comp.resolved_symbol, &a_variable);  // pre-resolved
+  }
+
+  // Submodule "n"
+  MUST_ASSIGN_LOOKUP_SYMBOL(module_n, module_m, "n");
+  EXPECT_EQ(module_n_info.metatype, SymbolMetaType::kModule);
+  EXPECT_EQ(module_n_info.file_origin, &src);
+  EXPECT_EQ(module_n_info.declared_type.syntax_origin, nullptr);
+
+  EXPECT_EQ(module_n_info.local_references_to_bind.size(), 1);
+  const auto n_ref_map(module_n_info.LocalReferencesMapViewForTesting());
+
+  ASSIGN_MUST_FIND(n_a_refs, n_ref_map, "a");
+  ASSERT_EQ(n_a_refs.size(), 1);  // references to "a" net in "n" module
+  for (const auto& n_a_ref : n_a_refs) {
+    const ReferenceComponent& n_a_ref_comp(n_a_ref->components->Value());
+    EXPECT_EQ(n_a_ref_comp.ref_type, ReferenceType::kUnqualified);
+    EXPECT_EQ(n_a_ref_comp.required_metatype, SymbolMetaType::kUnspecified);
+    EXPECT_EQ(n_a_ref_comp.identifier, "a");
+    EXPECT_EQ(n_a_ref_comp.resolved_symbol,
+              nullptr);  // resolving only in same scope
+  }
+
+  {
+    std::vector<absl::Status> resolve_diagnostics;
+    // Resolve mustn't break anything
+    symbol_table.Resolve(&resolve_diagnostics);
+    EXPECT_EMPTY_STATUSES(resolve_diagnostics);
+
+    // All references to "a" resolved.
+    for (const auto& a_ref : a_refs) {
+      const ReferenceComponent& a_ref_comp(a_ref->components->Value());
+      EXPECT_EQ(a_ref_comp.resolved_symbol,
+                &a_variable);  // resolved successfully
+    }
+
+    for (const auto& n_a_ref : n_a_refs) {
+      const ReferenceComponent& n_a_ref_comp(n_a_ref->components->Value());
+      EXPECT_EQ(n_a_ref_comp.resolved_symbol,
+                &a_variable);  // resolved successfully
```

Comment:
I might need a language lawyer to explain to me.  Is the second declaration of "a" in the nested module supposed to resolve to that in the outer module?  Maybe add a comment to this test case that cites the LRM passage about this.

Reply:
+1 for adding comment.